sigstore/dsse.py

```diff
@@ -210,6 +211,23 @@ def _from_json(cls, contents: bytes | str) -> Envelope:
         inner = _Envelope().from_json(contents)
         return cls(inner)
 
+    @classmethod
+    def from_payload(cls, payload_type: str, payload: bytes) -> Envelope:
```

Comment:
Like elsewhere: we should start with this being a private API, before we make it public.

```suggestion
    def _from_payload(cls, payload_type: str, payload: bytes) -> Envelope:
```

Reply:
It is the only "valid" way to create an envelope for users (except of just calling the private method).
Surely, I can change this but just like to get your thoughts on how otherwise a user can sign an envelope.

Reply:
Hmm, yeah. The problem here is that `from_payload` is a little misleading as an API: it produces an `Envelope`, but that `Envelope` is actually in an invalid state (since it's missing signatures).

Instead of passing an incomplete `Envelope` in for signing, maybe we could adapt the top-level `sign_dsse` signature:

```python
    def sign_dsse(
        self,
        input_: dsse.Statement | tuple[str, bytes],
    ) -> Bundle:
```

...where `tuple[str, bytes]` is the envelope type and arbitrary contents. That can then be passed into `dsse._sign` with some similar small tweaks.

(I still don't _love_ that signature, but I think it's a better starting point and avoids the need for a partial constructor here 🙂)

Reply:
How about using a dataclass?

```python
@dataclass
class DSSEPayload:
    payload_type: str
    payload: bytes
```

I think at some point the `sign_dsse` might need to be renamed :) but for now a dataclass would make it clearer for users.

WDYT?

Reply:
That works for me, although maybe `dsse.RawPayload` to make it clear that this is the "raw" variant of possible payload types (`dsse.Statement` being another valid variant).

Reply:
PTAL